### PR TITLE
[TT-8658] Changed optional fields of CRD to pointers

### LINKED ISSUE #606: [TT-8658] json serialization settings cause mis-alignment between desired and live manifests
When I create an api-definition setting some bool fields to their default values, the live manifest returned by kubernetes doesn't contain those fields. I am using ArgoCD to deploy my manifests through helm charts. The end result is that ArgoCD keeps complaining about out-of-sync api-definitions (see screenshot attached). Though currently I'm not experiencing the problem with strings, I'd expect the same problem.

## Expected Behavior
The live manifest should parse exactly like the desired manifest.

## Current Behavior
Boolean fields are excluded from the live manifest when they're set to `false`.

## Possible Solution
Replace `bool` fields with `*bool`. This will cause the serializer to output `false` when appropriate, and omit serialization only for non-set values.

## Steps to Reproduce
1. save the the following manifest in a file named `desired-test-api.yaml` and use it to create an api-definition by running `kubectl create -f desired-test-api.yaml`:
    ```yaml
    apiVersion: tyk.tyk.io/v1alpha1
    kind: ApiDefinition
    metadata:
      name: test-api
   spec:
      active: true
      proxy:
          strip_listen_path: false
          listen_path: /
    ```
3. save the live manifest into a temporary file: `kubectl get apidefinition test-api -o yaml > live-test-api.yaml`
4. compare the desired and live manifests: `diff desired-test-api.yaml live-test-api.yaml`. The `spec` object should be the same between the two files.

## Screenshots/Video
ArgoCD

## Your Environment

* OS (e.g. Linux, macOS, Windows):
  * OS version:
* `tyk-operator` version/commit:

### PR DESCRIPTION
**Pull request by komalsukhani**
## Description

Converted `string` and `bool` optional fields of APIDefinition and Security Policy to pointers.
Here is list of fields that are changed:
### API Definition Bool fields
	DisableHalfOpenState bool `json:"disable_half_open_state,omitempty"`
	Reverse      bool   `json:"reverse,omitempty"`
	UseExtendedPaths            bool              `json:"use_extended_paths,omitempty"`
	IgnoreEndpointCase          bool              `json:"ignore_endpoint_case,omitempty"`
	RequireSession bool   `json:"require_session,omitempty"`
	RawBodyOnly    bool   `json:"raw_body_only,omitempty"`
	EnableCache bool `json:"enable_cache,omitempty"`
	CacheAllSafeRequests bool `json:"cache_all_safe_requests,omitempty"`
	EnableUpstreamCacheControl bool `json:"enable_upstream_cache_control,omitempty"`
	Active bool `json:"active,omitempty"`
	EnableProxyProtocol bool `json:"enable_proxy_protocol,omitempty"`
	DoNotTrack *bool `json:"do_not_track,omitempty"`
	UseKeylessAccess *bool `json:"use_keyless,omitempty"`
	UseOauth2 bool `json:"use_oauth2,omitempty"`
	StripAuthData bool `json:"strip_auth_data,omitempty"`
	UseStandardAuth bool `json:"use_standard_auth,omitempty"`
	UseBasicAuth bool `json:"use_basic_auth,omitempty"`
	UseMutualTLSAuth      bool     `json:"use_mutual_tls_auth,omitempty"`
	EnableJWT bool `json:"enable_jwt,omitempty"`
	UseGoPluginAuth bool `json:"use_go_plugin_auth,omitempty"`
	EnableCoProcessAuth bool `json:"enable_coprocess_auth,omitempty"`
	JWTSkipKid bool `json:"jwt_skip_kid,omitempty"`
	DisableRateLimit bool `json:"disable_rate_limit,omitempty"`
	DisableQuota bool `json:"disable_quota,omitempty"`
	Internal bool `json:"internal,omitempty"`
	EnableIPWhiteListing bool `json:"enable_ip_whitelisting,omitempty"`
	EnableIPBlacklisting bool `json:"enable_ip_blacklisting,omitempty"`
	EnableContextVars bool `json:"enable_context_vars,omitempty"`
	EnableDetailedRecording bool `json:"enable_detailed_recording,omitempty"`
	PreserveHostHeader bool `json:"preserve_host_header,omitempty"`
	DisableStripSlash bool `json:"disable_strip_slash,omitempty"`
	StripListenPath bool `json:"strip_listen_path,omitempty"`
	EnableLoadBalancing bool `json:"enable_load_balancing,omitempty"`
	CheckHostAgainstUptimeTests bool `json:"check_host_against_uptime_tests,omitempty"`
	SSLInsecureSkipVerify bool `json:"ssl_insecure_skip_verify,omitempty"`
	SSLForceCommonNameCheck bool `json:"ssl_force_common_name_check,omitempty"`
	Enable bool `json:"enable,omitempty"`
	AllowCredentials bool `json:"allow_credentials,omitempty"`
	OptionsPassthrough bool `json:"options_passthrough,omitempty"`
	Debug bool `json:"debug,omitempty"`
	UseParam          bool            `json:"use_param,omitempty"`
	UseCookie         bool            `json:"use_cookie,omitempty"`
	UseCertificate    bool            `json:"use_certificate,omitempty"`
	ValidateSignature bool            `json:"validate_signature,omitempty"`
	DisableQueryBatching bool                    `json:"disable_query_batching,omitempty"`
	Compressed bool   `json:"compressed,omitempty"`
	Graph      bool   `json:"graph,omitempty"`

### API Definition String fields
	RewriteTo         string                `json:"rewrite_to,omitempty"`
	RewriteTo string `json:"rewrite_to,omitempty"`
	Path string `json:"path,omitempty"`
	Query string `json:"query,omitempty"`
	Path string `json:"path,omitempty"`
	Query string `json:"query,omitempty"`
	Namespace string `json:"namespace,omitempty"`
	Expires                     string            `json:"expires,omitempty"`
	OverrideTarget              string            `json:"override_target,omitempty"`
	HeaderName      string `json:"header_name,omitempty"`
	FormParamName   string `json:"param_name,omitempty"`
	RegexExpression string `json:"regex_expression,omitempty"`
	CacheControlTTLHeader string `json:"cache_control_ttl_header,omitempty"`
	ID string `json:"id,omitempty" hash:"ignore"`
	APIID string `json:"api_id,omitempty"`
	OrgID string `json:"org_id,omitempty"`
	Domain string `json:"domain,omitempty"`
	JWTSigningMethod string `json:"jwt_signing_method,omitempty"`
	JWTSource string `json:"jwt_source,omitempty"`
	JWTIdentityBaseField string `json:"jwt_identity_base_field,omitempty"`
	JWTClientIDBaseField string `json:"jwt_client_base_field,omitempty"`
	JWTPolicyFieldName string `json:"jwt_policy_field_name,omitempty"`
	JWTScopeClaimName string `json:"jwt_scope_claim_name,omitempty"`
	CustomMiddlewareBundle string            `json:"custom_middleware_bundle,omitempty"`
	ListenPath string `json:"listen_path,omitempty"`
	ProxyURL string `json:"proxy_url,omitempty"`
	AuthLoginRedirect string `json:"auth_login_redirect,omitempty"`
	ParamName         string          `json:"param_name,omitempty"`
	CookieName        string          `json:"cookie_name,omitempty"`
	MergedSDL            string                  `json:"merged_sdl,omitempty"`
	Schema string `json:"schema,omitempty"`
	GraphRef string `json:"graph_ref,omitempty"`
	Body                       string                      `json:"body,omitempty"`
	DefaultTypeName            string                      `json:"default_type_name,omitempty"`
	TypeName   string `json:"type_name,omitempty"`
	Query      string `json:"q,omitempty"`
	Sort       string `json:"sort,omitempty"`
	Category   string `json:"category,omitempty"`
	AuthType   string `json:"auth_type,omitempty"`
	
### Security Policy Bool fields
	IsInactive        bool                        `json:"is_inactive,omitempty"`
	HMACEnabled                   bool `json:"hmac_enabled,omitempty"`
	EnableHTTPSignatureValidation bool `json:"enable_http_signature_validation,omitempty"`
	DisableIntrospection bool `json:"disable_introspection,omitempty"`
	Quota      bool `json:"quota,omitempty"`
	RateLimit  bool `json:"rate_limit,omitempty"`
	Complexity bool `json:"complexity,omitempty"`
	Acl        bool `json:"acl,omitempty"`
	PerAPI     bool `json:"per_api,omitempty"`
	
### Security Policy String fields
	MID string `json:"_id,omitempty" oss:"ignore"`
	ID string `json:"id,omitempty"`
	OrgID string `json:"org_id,omitempty"`
	APIName string `json:"api_name,omitempty"`
	APIID    string   `json:"api_id,omitempty"`
	AllowanceScope string       `json:"allowance_scope,omitempty"`
	TypeName string `json:"type_name,omitempty"`
	FieldName string `json:"field_name,omitempty"`

## Related Issue

[TT-8658](https://tyktech.atlassian.net/browse/TT-8818)
Resolves #606 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [x] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [x] Make sure you are updating [CHANGELOG.md](https://github.com/TykTechnologies/tyk-operator/blob/master/CHANGELOG.md) based on your changes.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] If you've changed API models, please update CRDs.
  - [ ] `make manifests`
  - [ ] `make helm`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `gofmt -s -w .`
  - [x] `go vet ./...`
  - [x] `golangci-lint run`

[TT-8658]: https://tyktech.atlassian.net/browse/TT-8658?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by andrei-tyk — APPROVED** [review]

File: controllers/common.go [hunk 1]
```diff
@@ -291,8 +296,13 @@ func updateOperatorContextStatus(
...
 
 		var operatorContext v1alpha1.OperatorContext
+		namespace := ""
+
+		if ctxRef.Namespace != nil {
+			namespace = *ctxRef.Namespace
+		}
 
-		key := types.NamespacedName{Name: ctxRef.Name, Namespace: ctxRef.Namespace}
+		key := types.NamespacedName{Name: ctxRef.Name, Namespace: namespace}
```
Comment (buraksekili) [inline controllers/common.go:305]: Is providing empty namespace safe for k8s client?

File: pkg/snapshot/snapshot.go [hunk 2]
```diff
@@ -402,11 +407,20 @@ func writePolicy(idx int, userPolicy *tykv1alpha1.SecurityPolicySpec, w *bufio.W
...
+
+	if pol.Spec.OrgID == nil {
+		pol.Spec.OrgID = new(string)
+	}
+
+	*pol.Spec.OrgID = ""
 
 	for i := 0; i < len(pol.Spec.AccessRightsArray); i++ {
-		apiID := pol.Spec.AccessRightsArray[i].APIID
+		apiID := *pol.Spec.AccessRightsArray[i].APIID
```
Comment (buraksekili) [inline pkg/snapshot/snapshot.go:423]: what do you think checking `pol.Spec.AccessRightsArray[i]` is nil or not before accessing APIID?

File: integration/securitypolicy_test.go [hunk 3]
```diff
@@ -956,7 +981,9 @@ func TestSecurityPolicyWithContextRef(t *testing.T) {
 
 						linkedPolMeta := oc.Status.LinkedSecurityPolicies[0]
 
-						return linkedPolMeta.Name == policy.Name && linkedPolMeta.Namespace == policy.Namespace
+						return linkedPolMeta.Name == policy.Name &&
+							linkedPolMeta.Namespace != nil &&
+							*linkedPolMeta.Namespace == policy.Namespace
```
Comment (buraksekili) [inline integration/securitypolicy_test.go:986]: can it cause a panic if `linkedPolMeta.Namespace` is nil?

File: controllers/ingress_controller.go [hunk 4]
```diff
@@ -121,13 +121,15 @@ func (r *IngressReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+	useKeyless := true
+
 	return &v1alpha1.ApiDefinition{
 		Spec: v1alpha1.APIDefinitionSpec{
 			APIDefinitionSpec: model.APIDefinitionSpec{
 				Name:             "default-keyless",
 				Protocol:         "http",
-				UseKeylessAccess: true,
-				Active:           true,
+				UseKeylessAccess: &useKeyless,
```
Comment (buraksekili) [inline controllers/ingress_controller.go:131]: Is it safe to use same variable's address for two fields? I mean, if `UseKeylessAccess`'s value is changed to false, it'll affect `Active` as well

File: controllers/common.go (hunk 1 above)
Comment (komalsukhani) [reply controllers/common.go:305]: Doesn't it use `default` namespace when not set?

File: controllers/ingress_controller.go (hunk 4 above)
Comment (komalsukhani) [reply controllers/ingress_controller.go:131]: Opps sorry. 
Didn't intend to do it that way. Will fix it

File: pkg/client/dashboard/security_policy.go [hunk 5]
```diff
@@ -79,7 +84,11 @@ func (p SecurityPolicy) Create(ctx context.Context, def *v1.SecurityPolicySpec)
 
 // Update updates a resource object def
 func (p SecurityPolicy) Update(ctx context.Context, def *v1.SecurityPolicySpec) error {
-	res, err := client.PutJSON(ctx, client.Join(endpointPolicies, def.MID), def)
+	if def.MID == nil || *def.MID == "" {
+		return fmt.Errorf("Failed to update policy. Missing policy id")
```
[the remark this answers is not in the document]
Comment (komalsukhani) [reply pkg/client/dashboard/security_policy.go]: Fixed in [1acdb15](https://github.com/TykTechnologies/tyk-operator/pull/627/commits/1acdb154f3cccc0675ecf005fa863923882f4cc0)

File: controllers/apidefinition_controller.go [hunk 6]
```diff
@@ -948,36 +964,45 @@ func (r *ApiDefinitionReconciler) processSubGraphExec(ctx context.Context, urs *
...
 			Name:      ref.Name,
-			Namespace: ns,
+			Namespace: *ns,
 		}, subGraph)
 		if err != nil {
 			return err
 		}
 
-		ns, name := decodeID(subGraph.Status.LinkedByAPI)
+		api_ns, api_name := decodeID(subGraph.Status.LinkedByAPI)
```
Comment (buraksekili) [inline controllers/apidefinition_controller.go:1002]: not crucial but can we use camelCase here?

**Review by buraksekili — APPROVED** [review]
🚀 thanks @komalsukhani, lgtm! 🤞

File: controllers/common.go (hunk 1 above)
Comment (komalsukhani) [reply controllers/common.go:305]: Set default value to `default` instead of empty

Comment (asoorm) [reply controllers/common.go:305]: If not set, I would not explicitly set to default. I would inherit the namespace from the context.

Comment (caroltyk) [reply controllers/common.go:305]: Let's fix this in another PR

**Pull request merged**